### PR TITLE
Make fish build more quietly.

### PR DESCRIPTION
**Pull request by floam**
## Description

I made `make` a lot quieter and want to get feedback. It needs to @echo a few more things than it does now and it makes travis mad. It also currently lacks a way to flip it off aside from commenting out the added definition at the top of Makefile.in. If people generally are cool with it, that should all be pretty trivial. 

I think we should be more likely to notice warning output now, with a better SNR. 

## Output

<details><summary>Here's the new output (hit triangle).</summary>
```
FISH_BUILD_VERSION = 2.3.1-393-g819cf26
clang++ [src/complete.cpp]
In file included from src/complete.cpp:35:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/event.cpp]
In file included from src/event.cpp:16:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/exec.cpp]
In file included from src/exec.cpp:37:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/expand.cpp]
clang++ [src/fallback.cpp]
clang++ [src/fish_version.cpp]
clang++ [src/function.cpp]
clang++ [src/highlight.cpp]
clang++ [src/history.cpp]
clang++ [src/input.cpp]
In file included from src/input.cpp:33:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/input_common.cpp]
clang++ [src/intern.cpp]
clang++ [src/io.cpp]
clang++ [src/iothread.cpp]
clang++ [src/kill.cpp]
clang++ [src/output.cpp]
clang++ [src/pager.cpp]
clang++ [src/parse_execution.cpp]
In file included from src/parse_execution.cpp:39:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/parse_productions.cpp]
clang++ [src/parse_tree.cpp]
clang++ [src/parse_util.cpp]
clang++ [src/parser.cpp]
In file included from src/parser.cpp:21:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
src/parser.cpp:258:10: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
block_t *const parser_t::current_block() { return block_stack.empty() ? NULL : block_stack.back(); }
         ^~~~~~
2 warnings generated.
clang++ [src/parser_keywords.cpp]
clang++ [src/path.cpp]
clang++ [src/postfork.cpp]
clang++ [src/proc.cpp]
In file included from src/proc.cpp:41:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/reader.cpp]
In file included from src/reader.cpp:64:
./src/parser.h:274:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
    block_t *const current_block();
             ^~~~~~
1 warning generated.
clang++ [src/sanity.cpp]
clang++ [src/screen.cpp]
clang++ [src/signal.cpp]
clang++ [src/tokenizer.cpp]
clang++ [src/utf8.cpp]
clang++ [src/util.cpp]
clang++ [src/wcstringutil.cpp]
clang++ [src/wgetopt.cpp]
clang++ [src/wildcard.cpp]
clang++ [src/wutil.cpp]

Compiling fish with CXXFLAGS=-g -O2 -Wignored-qualifiers -fno-exceptions -Wall -Wno-sign-compare -I/usr/local/Cellar/pcre2/10.21/include -iquote. -iquote./src/ -DLOCALEDIR="/usr/local/share/locale" -DPREFIX=L"/usr/local" -DDATADIR=L"/usr/local/share" -DSYSCONFDIR=L"/usr/local/etc" -DBINDIR=L"/usr/local/bin" -DDOCDIR=L"/usr/local/share/doc/fish" -DFISH_BUILD_VERSION="2.3.1-393-g819cf26"
===
clang++ [src/fish_indent.cpp]
clang++ [src/print_help.cpp]

Compiling fish_indent with CXXFLAGS=-g -O2 -Wignored-qualifiers -fno-exceptions -Wall -Wno-sign-compare -I/usr/local/Cellar/pcre2/10.21/include -iquote. -iquote./src/ -DLOCALEDIR="/usr/local/share/locale" -DPREFIX=L"/usr/local" -DDATADIR=L"/usr/local/share" -DSYSCONFDIR=L"/usr/local/etc" -DBINDIR=L"/usr/local/bin" -DDOCDIR=L"/usr/local/share/doc/fish" -DFISH_BUILD_VERSION="2.3.1-393-g819cf26"

clang++ [src/fish_key_reader.cpp]
src/fish_key_reader.cpp:55:14: warning: 'const' type qualifier on return type has no effect [-Wignored-qualifiers]
static char *const sequence_name(wchar_t wc) {
             ^~~~~~
1 warning generated.

Compiling fish_key_reader with CXXFLAGS=-g -O2 -Wignored-qualifiers -fno-exceptions -Wall -Wno-sign-compare -I/usr/local/Cellar/pcre2/10.21/include -iquote. -iquote./src/ -DLOCALEDIR="/usr/local/share/locale" -DPREFIX=L"/usr/local" -DDATADIR=L"/usr/local/share" -DSYSCONFDIR=L"/usr/local/etc" -DBINDIR=L"/usr/local/bin" -DDOCDIR=L"/usr/local/share/doc/fish" -DFISH_BUILD_VERSION="2.3.1-393-g819cf26"
/Users/floam/src/githubfish/doc.h:5646: warning: found </pre> tag without matching <pre>
/Users/floam/src/githubfish/doc.h:5646: warning: unexpected command endhtmlonly
Lexicon filter is not available. Continuing without.
fish has now been built.
Use '/Applications/Xcode-beta.app/Contents/Developer/usr/bin/make install' to install fish.
install: share/tools/web_config/js: Inappropriate file type or format
install: share/tools/web_config/partials: Inappropriate file type or format
install: share/tools/web_config/sample_prompts: Inappropriate file type or format
fish is now installed on your system.
To run fish, type 'fish' in your terminal.

To use fish as your login shell:
* use the command 'chsh -s /usr/local/bin/fish'.

To set your colors, run 'fish_config'
To scan your man pages for completions, run 'fish_update_completions'
To autocomplete command suggestions press Ctrl + F or right arrow key.

Have fun!
```
</details>
 (I added `-Wignored-qualifiers` to just make it a little more noisy as fish currently builds with 0 warnings.)

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -679,7 +679,9 @@ obj:
 # Build the fish program.
 #
 fish: obj/fish.o $(FISH_OBJS) $(EXTRA_PCRE2)
+	echo
 	echo Compiling fish with CXXFLAGS=$(CXXFLAGS)
+	echo ===
```
Comment (krader1961) [inline Makefile.in]: The "===" is inconsistent with similar echo sequences. Personally I prefer surrounding meta comments with "====" (though I typically use ten or twenty equal-signs) so I have no objection to your including such a change but we should be consistent.

File: Makefile.in [hunk 2]
```diff
@@ -513,167 +515,131 @@ endif
...
-	done;
+	$(foreach i, $(wildcard user_doc/html/*), $(INSTALL) -m 644 $(i) $(DESTDIR)$(docdir);)
+	$(INSTALL) -m 644 CHANGELOG.md $(DESTDIR)$(docdir)
+	
 	$(INSTALL) -m 755 -d $(DESTDIR)$(mandir)/man1
-	for i in $(MANUALS); do \
-		$(INSTALL) -m 644 $$i $(DESTDIR)$(mandir)/man1/; \
-		true; \
-	done;
+	$(foreach i, $(MANUALS), $(INSTALL) -m 644 $(i) $(DESTDIR)$(mandir)/man1/;)
```
Comment (krader1961) [inline Makefile.in]: I'm guessing this change is the reason Travis is failing. Perhaps revert this portion.

File: Makefile.in (hunk 1 above)
Comment (floam) [inline Makefile.in]: Whoops, I thought I put `===` on every header. I updated it as you suggested. .

File: Makefile.in [hunk 3]
```diff
@@ -242,13 +245,12 @@ prof: all
 #
 # User documentation, describing the features of the fish shell.
 #
-# Depend on the sources (*.hdr.in) and manually make the intermediate *.hdr
```
Comment (zanchey) [inline Makefile.in]: ??

**Comment by krader1961** [comment]
I love the idea and don't have any objections once you fix the Travis failures and the minor style inconsistency I noted. I would love to be able to "make fish" and not have to wade through all the individual commands in order to see the warnings and errors. I wouldn't mind seeing the `-Wignored-qualifiers` addition be made permanent.

**Comment by zanchey** [comment]
This sounds like a good idea, and I tried to do something similar a few years ago.

There's already a quasi-standard for this sort of approach - [Automake silent rules](https://www.gnu.org/software/automake/manual/html_node/Automake-Silent-Rules.html) & its pure-Make implementation in Linux's Kbuild.

In particular, I think it would be helpful if `make V=1` or similar turned the full printing of rules back on - I use this quite a lot to get the full commandline and then make small modifications such as using `-E`.

File: Makefile.in (hunk 2 above)
Comment (floam) [inline Makefile.in]: It was something else.  It's working now.

File: Makefile.in (hunk 3 above)
Comment (floam) [inline Makefile.in]: I don't know, whoops!

**Comment by floam** [comment]
I see now that probably I should also the same output that you see there and that PCRE2 is using - I hadn't tried a build that would cause that to be built at the same time but these should flow together.

```
clang++ [src/util.cpp]
clang++ [src/wcstringutil.cpp]
clang++ [src/wgetopt.cpp]
clang++ [src/wildcard.cpp]
clang++ [src/wutil.cpp]
  CC       src/libpcre2_32_la-pcre2_auto_possess.lo
  CC       src/libpcre2_32_la-pcre2_compile.lo
  CC       src/libpcre2_32_la-pcre2_config.lo
  CC       src/libpcre2_32_la-pcre2_context.lo
  CC       src/libpcre2_32_la-pcre2_dfa_match.lo
```

**Comment by krader1961** [comment]
@zanchey I'm not thrilled about the solution documented in the link you provided. I'd prefer that a silent build be the default without having to provide a flag to the build tools. I do recognize that sometimes you need to see the actual commands being run. Such as when I implemented the `make lint` target. But I think it's sufficient to be able to just edit the *Makefile* and change the one line that makes the build silent. That, however, should be added to the *CONTRIBUTING.md* file.

**Comment by zanchey** [comment]
We can make the silent build be the default, that's what the Linux Kbuild stuff does.

**Comment by floam** [comment]
Here's what the PR is currently displaying like:

screenshot 2016-07-18 at 3 21 55 am

**Comment by floam** [comment]
I can't figure out how on earth the current commit is not making a `fish_indent`.

**Comment by floam** [comment]
Do we want to merge this? Or try another approach?

**Pull request merged**